storage/store.go

```diff
@@ -1701,6 +1707,42 @@ func (s *Store) MergeRange(subsumingRng *Replica, updatedEndKey roachpb.RKey, su
 	return subsumingRng.setDesc(&copy)
 }
 
+// If the subsuming replica has the range lease, then we update its timestamp
+// cache with the entries from the subsumed. If it's not the lease holder,
+// then the timestamp cache doesn't matter (in fact it should be empty, to
+// save memory).
+func (s *Store) maybeMergeTimestampCaches(
+	subsumingRng *Replica,
+	subsumedRng *Replica,
+) error {
+	subsumingRng.mu.Lock()
+	defer subsumingRng.mu.Unlock()
+	subsumingLease := subsumingRng.mu.state.Lease
+
+	subsumedRng.mu.Lock()
+	defer subsumedRng.mu.Unlock()
+	subsumedLease := subsumedRng.mu.state.Lease
+
+	// TODO(andrei): I think checking the subsumed leases like this is incorrect,
+	// since the stores of the different replicas of subsuming can have different
+	// state for subsumed? Unless the subsumed range has been quiesced somehow.
+	if subsumedLease.Covers(s.Clock().Now()) &&
+		subsumingLease.Replica.StoreID != subsumedLease.Replica.StoreID {
+		// TODO(andrei): What should we do if we're not the lease holder for
```

Comment:
Freezing the RHS in preparation for a merge should also terminate all leases, so there will be no active lease holder on the RHS during the merge. For now, returning an error if the leases don't match is probably the best we can do.